ibis/pandas/client.py

```diff
@@ -163,29 +168,72 @@ def ibis_dtype_to_pandas(ibis_dtype):
     if isinstance(ibis_dtype, dt.Timestamp) and ibis_dtype.timezone:
         return DatetimeTZDtype('ns', ibis_dtype.timezone)
     elif isinstance(ibis_dtype, dt.Interval):
-        return 'timedelta64[{}]'.format(ibis_dtype.unit)
+        return np.dtype('timedelta64[{}]'.format(ibis_dtype.unit))
+    elif isinstance(ibis_dtype, dt.Category):
+        return CategoricalDtype()
     elif type(ibis_dtype) in _ibis_dtypes:
         return _ibis_dtypes[type(ibis_dtype)]
     else:
-        return 'object'
+        return np.dtype(np.object_)
 
 
 def ibis_schema_to_pandas(schema):
     return list(zip(schema.names, map(ibis_dtype_to_pandas, schema.types)))
 
 
 def ibis_schema_apply_to(schema, df):
-    """Applies the Ibis schema on a pandas dataframe"""
+    """Applies the Ibis schema to a pandas DataFrame
+
+    Parameters
+    ----------
+    schema : ibis.schema.Schema
+    df : pandas.DataFrame
+
+    Returns
+    -------
+    df : pandas.DataFrame
+
+    Notes
+    -----
+    Mutates `df`
+    """
 
     for column, dtype in schema.items():
         pandas_dtype = dtype.to_pandas()
-        if isinstance(dtype, dt.Interval):
-            df[column] = df[column].values.astype(pandas_dtype)
-        else:
-            df[column] = df[column].astype(pandas_dtype, errors='ignore')
-
-        if PY2 and dtype == dt.string:
-            df[column] = df[column].str.decode('utf-8', errors='ignore')
+        col = df[column]
+        col_dtype = col.dtype
+
+        try:
+            not_equal = pandas_dtype != col_dtype
+        except TypeError:  # ugh, we can't compare dtypes coming from pandas
+            not_equal = False
+
+        if not_equal:
+            if isinstance(dtype, dt.Interval):
+                df[column] = col.values.astype(pandas_dtype)
+            elif isinstance(dtype, dt.Timestamp):
+                # input column has a timezone
+                timezone = dtype.timezone
```

Comment:
Uhh, this seems complex, there are 5 possible outcomes in case of timestamps. 
How about factoring this out with slightly more comments?